File: app/src/main/java/dev/ragnarok/fenrir/fragment/search/AbsSearchFragment.java
package dev.ragnarok.fenrir.fragment.search;

import static dev.ragnarok.fenrir.util.Objects.nonNull;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.StringRes;
import androidx.recyclerview.widget.RecyclerView;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import dev.ragnarok.fenrir.R;
import dev.ragnarok.fenrir.fragment.base.PlaceSupportMvpFragment;
import dev.ragnarok.fenrir.fragment.search.options.BaseOption;
import dev.ragnarok.fenrir.listener.EndlessRecyclerOnScrollListener;
import dev.ragnarok.fenrir.mvp.presenter.search.AbsSearchPresenter;
import dev.ragnarok.fenrir.mvp.view.search.IBaseSearchView;
import dev.ragnarok.fenrir.util.ViewUtils;

public abstract class AbsSearchFragment<P extends AbsSearchPresenter<V, ?, T, ?>, V extends IBaseSearchView<T>, T, A extends RecyclerView.Adapter<?>>
        extends PlaceSupportMvpFragment<P, V> implements IBaseSearchView<T> {
    public A mAdapter;
    private SwipeRefreshLayout mSwipeRefreshLayout;
    private TextView mEmptyText;

    private void onSeachOptionsChanged() {
        callPresenter(p -> p.fireOptionsChanged());
    }

    public View createViewLayout(@NonNull LayoutInflater inflater, @Nullable ViewGroup container) {
        return inflater.inflate(R.layout.fragment_search, container, false);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View root = createViewLayout(inflater, container);

        RecyclerView recyclerView = root.findViewById(R.id.list);
        RecyclerView.LayoutManager manager = createLayoutManager();
        recyclerView.setLayoutManager(manager);
        recyclerView.addOnScrollListener(new EndlessRecyclerOnScrollListener() {
            @Override
            public void onScrollToLastElement() {
                callPresenter(p -> p.fireScrollToEnd());
            }
        });

        mAdapter = createAdapter(Collections.emptyList());
        recyclerView.setAdapter(mAdapter);

        mSwipeRefreshLayout = root.findViewById(R.id.refresh);
        mSwipeRefreshLayout.setOnRefreshListener(() -> callPresenter(p -> p.fireRefresh()));
        ViewUtils.setupSwipeRefreshLayoutWithCurrentTheme(requireActivity(), mSwipeRefreshLayout);

        mEmptyText = root.findViewById(R.id.empty);
        mEmptyText.setText(getEmptyText());
        postCreate(root);
        return root;
    }

    @StringRes
    int getEmptyText() {
        return R.string.list_is_empty;
    }

    public void fireTextQueryEdit(String q) {
        callPresenter(p -> p.fireTextQueryEdit(q));
    }

    @Override
    public void displayData(List<T> data) {
        if (nonNull(mAdapter)) {
            setAdapterData(mAdapter, data);
        }
    }

    @Override
    public void notifyItemChanged(int index) {
        if (nonNull(mAdapter)) {
            mAdapter.notifyItemChanged(index);
        }
    }

    @Override
    public void setEmptyTextVisible(boolean visible) {
        if (nonNull(mEmptyText)) {
            mEmptyText.setVisibility(visible ? View.VISIBLE : View.GONE);
        }
    }

    @Override
    public void notifyDataSetChanged() {
        if (nonNull(mAdapter)) {
            mAdapter.notifyDataSetChanged();
        }
    }

    @Override
    public void notifyDataAdded(int position, int count) {
        if (nonNull(mAdapter)) {
            mAdapter.notifyItemRangeInserted(position, count);
        }
    }

    @Override
    public void showLoading(boolean loading) {
        if (nonNull(mSwipeRefreshLayout)) {
            mSwipeRefreshLayout.setRefreshing(loading);
        }
    }

    public void openSearchFilter() {
        callPresenter(p -> p.fireOpenFilterClick());
    }

    @Override
    public void displayFilter(int accountId, ArrayList<BaseOption> options) {
        FilterEditFragment fragment = FilterEditFragment.newInstance(accountId, options);
        getParentFragmentManager().setFragmentResultListener(FilterEditFragment.REQUEST_FILTER_EDIT, fragment, (requestKey, result) -> onSeachOptionsChanged());
        fragment.show(getParentFragmentManager(), "filter-edit");
    }

    abstract void setAdapterData(A adapter, List<T> data);

    abstract void postCreate(View root);

    abstract A createAdapter(List<T> data);

    abstract RecyclerView.LayoutManager createLayoutManager();
}

File: app/src/main/java/dev/ragnarok/fenrir/fragment/search/AbsSearchFragment.kt
package dev.ragnarok.fenrir.fragment.search

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.annotation.StringRes
import androidx.recyclerview.widget.RecyclerView
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import dev.ragnarok.fenrir.R
import dev.ragnarok.fenrir.fragment.base.PlaceSupportMvpFragment
import dev.ragnarok.fenrir.fragment.search.options.BaseOption
import dev.ragnarok.fenrir.listener.EndlessRecyclerOnScrollListener
import dev.ragnarok.fenrir.mvp.presenter.search.AbsSearchPresenter
import dev.ragnarok.fenrir.mvp.view.search.IBaseSearchView
import dev.ragnarok.fenrir.util.ViewUtils.setupSwipeRefreshLayoutWithCurrentTheme

abstract class AbsSearchFragment<P : AbsSearchPresenter<V, *, T, *>, V : IBaseSearchView<T>, T, A : RecyclerView.Adapter<*>> :
    PlaceSupportMvpFragment<P, V>(), IBaseSearchView<T> {
    @JvmField
    var mAdapter: A? = null
    private var mSwipeRefreshLayout: SwipeRefreshLayout? = null
    private var mEmptyText: TextView? = null
    protected var recyclerView: RecyclerView? = null
    private fun onSearchOptionsChanged() {
        presenter?.fireOptionsChanged()
    }

    open fun createViewLayout(inflater: LayoutInflater, container: ViewGroup?): View {
        return inflater.inflate(R.layout.fragment_search, container, false)
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        val root = createViewLayout(inflater, container)
        recyclerView = root.findViewById(R.id.list)
        val manager = createLayoutManager()
        recyclerView?.layoutManager = manager
        recyclerView?.addOnScrollListener(object : EndlessRecyclerOnScrollListener() {
            override fun onScrollToLastElement() {
                presenter?.fireScrollToEnd()
            }
        })
        mAdapter = createAdapter(mutableListOf())
        recyclerView?.adapter = mAdapter
        mSwipeRefreshLayout = root.findViewById(R.id.refresh)
        mSwipeRefreshLayout?.setOnRefreshListener {
            presenter?.fireRefresh()
        }
        setupSwipeRefreshLayoutWithCurrentTheme(requireActivity(), mSwipeRefreshLayout)
        mEmptyText = root.findViewById(R.id.empty)
        mEmptyText?.setText(emptyText)
        postCreate(root)
        return root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        parentFragmentManager.setFragmentResultListener(
            FilterEditFragment.REQUEST_FILTER_EDIT,
            this
        ) { _: String?, _: Bundle? -> onSearchOptionsChanged() }
    }

    @get:StringRes
    val emptyText: Int
        get() = R.string.list_is_empty

    fun fireTextQueryEdit(q: String?) {
        presenter?.fireTextQueryEdit(q)
    }

    override fun displayData(data: MutableList<T>) {
        mAdapter?.let {
            setAdapterData(it, data)
        }
    }

    override fun notifyItemChanged(index: Int) {
        mAdapter?.notifyItemChanged(index)
    }

    override fun setEmptyTextVisible(visible: Boolean) {
        mEmptyText?.visibility = if (visible) View.VISIBLE else View.GONE
    }

    override fun notifyDataSetChanged() {
        mAdapter?.notifyDataSetChanged()
    }

    override fun notifyDataAdded(position: Int, count: Int) {
        mAdapter?.notifyItemRangeInserted(position, count)
    }

    override fun showLoading(loading: Boolean) {
        mSwipeRefreshLayout?.isRefreshing = loading
    }

    fun openSearchFilter() {
        presenter?.fireOpenFilterClick()
    }

    override fun displayFilter(accountId: Int, options: ArrayList<BaseOption>) {
        val fragment = FilterEditFragment.newInstance(accountId, options)
        fragment.show(parentFragmentManager, "filter-edit")
    }

    abstract fun setAdapterData(adapter: A, data: MutableList<T>)
    abstract fun postCreate(root: View)
    abstract fun createAdapter(data: MutableList<T>): A
    abstract fun createLayoutManager(): RecyclerView.LayoutManager
}